v03_pipeline/lib/misc/pedigree.py

```diff
@@ -1,28 +1,176 @@
+@dataclass
+class Sample:
+    sample_id: str
+    sex: Ploidy
+    mother: str = None
+    father: str = None
+    maternal_grandmother: str = None
+    maternal_grandfather: str = None
+    paternal_grandmother: str = None
+    paternal_grandfather: str = None
+    siblings: list[str] = field(default_factory=list)
+    half_siblings: list[str] = field(default_factory=list)
+    aunt_nephews: list[str] = field(default_factory=list)
+
+    def is_aunt_nephew(self: Sample, other: Sample) -> bool:
+        return (
+            # My Maternal Grandparents are your Parents
+            self.maternal_grandmother
+            and self.maternal_grandfather
+            and (self.maternal_grandmother == other.mother)
+            and (self.maternal_grandfather == other.father)
+        ) or (
+            # My Paternal Grandparents are your Parents
+            self.paternal_grandmother
+            and self.paternal_grandfather
+            and (self.paternal_grandmother == other.mother)
+            and (self.paternal_grandfather == other.father)
+        )
+
+
+@dataclass
+class Family:
+    family_guid: str
+    samples: dict[str, Sample]
+
+    def __hash__(self):
+        return hash(self.family_guid)
+
+    @staticmethod
+    def parse_direct_lineage(rows: list[hl.Struct]) -> dict[str, Sample]:
+        samples = {}
+        for row in rows:
+            samples[row.s] = Sample(
+                sample_id=row.s,
+                sex=Ploidy(row.sex),
+                mother=row.maternal_s,
+                father=row.paternal_s,
+            )
+
+        for row in rows:
+            # Maternal GrandParents
+            maternal_s = samples[row.s].mother
+            if maternal_s and samples[maternal_s].mother:
+                samples[row.s].maternal_grandmother = samples[maternal_s].mother
+            if maternal_s and samples[maternal_s].father:
+                samples[row.s].maternal_grandfather = samples[maternal_s].father
+
+            # Paternal GrandParents
+            paternal_s = samples[row.s].father
+            if paternal_s and samples[paternal_s].mother:
+                samples[row.s].paternal_grandmother = samples[paternal_s].mother
+            if paternal_s and samples[paternal_s].father:
+                samples[row.s].paternal_grandfather = samples[paternal_s].father
+        return samples
+
+    @staticmethod
+    def parse_collateral_lineage(
+        samples: dict[str, Sample],
+    ) -> dict[str, Sample]:
+        # NB: relationships are identified unidirectionally here (for better or for worse)
+        # A sample_i that is siblings with sample_j, will list sample_j as as sibling, but
+        # sample_j will not list sample_i as a sibling.  Relationships only appear in the
+        # ibd table a single time, so we only need to check the pairing once.
+        for sample_i, sample_j in itertools.combinations(samples.keys(), 2):
+            # If other sample is already related, continue
+            if sample_j in {
```

Comment:
not super important, but you use a series of `if` with a `continue` at the end, and that could be rewritten as `elif`